In [114]:
# Dependencies and Setup
import hvplot.pandas
import pandas as pd
import requests
from markupsafe import Markup #Caution: this may need to be updated in the condas env being used

# Import API key
from api_keys import geoapify_key

In [115]:
# Load the SF_Crime_Data_from_2018_to_Present.csv CSV file into a Pandas DataFrame
city_data_df = pd.read_csv("SF_Crime_Data_from_2018_to_Present.csv")

# Display sample data
city_data_df.tail()

Incident Datetime Incident Date Incident Time  Incident Year  \
783433       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783434       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783435       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783436       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783437       1/1/18 0:00        1/1/18   12:00:00 AM           2018   

       Incident Day of Week Report Datetime        Row ID  Incident ID  \
783433               Monday  11/20/20 14:21  9.803191e+10       980319   
783434               Monday   1/23/21 12:45  9.987897e+10       998789   
783435               Monday   5/28/21 12:57  1.030000e+11      1034762   
783436               Monday    5/4/21 18:13  1.030000e+11      1030988   
783437               Monday   8/10/21 13:10  1.060000e+11      1058884   

        Incident Number   CAD Number  ... Police District  \
783433        200701703  203251939.0  ...        Northern   
783434        210051968  210231480.0  ...       Out of SF   
783435        210329436  211481637.0  ...         Taraval   
783436        216055213          NaN  ...        Southern   
783437        210508088  212221621.0  ...      Tenderloin   

             Analysis Neighborhood  Supervisor District   Latitude  \
783433                      Marina                  2.0  37.803727   
783434       Bayview Hunters Point                 10.0  37.729910   
783435  Oceanview/Merced/Ingleside                 11.0  37.721659   
783436                         NaN                  NaN        NaN   
783437                  Tenderloin                  5.0  37.783933   

         Longitude                                          Point  \
783433 -122.428400  POINT (-122.42840027589088 37.80372712877811)   
783434 -122.397177  POINT (-122.39717687851248 37.72990988256753)   
783435 -122.467615  POINT (-122.46761532399763 37.72165880704187)   
783436         NaN                                            NaN   
783437 -122.412595   POINT (-122.41259527758581 37.7839325760642)   

       Civic Center Harm Reduction Project Boundary  \
783433                                          NaN   
783434                                          NaN   
783435                                          NaN   
783436                                          NaN   
783437                                          1.0   

       HSOC Zones as of 2018-06-05 Current Supervisor Districts  \
783433                         NaN                          6.0   
783434                         NaN                          9.0   
783435                         NaN                          1.0   
783436                         NaN                          NaN   
783437                         NaN                         10.0   

       Current Police Districts  
783433                      4.0  
783434                      2.0  
783435                     10.0  
783436                      NaN  
783437                      5.0  

[5 rows x 28 columns]

In [116]:
# Count NaN values in each column of city_data_df
nan_count_per_column = city_data_df.isna().sum()

# Count total NaN values in the entire city_data_df DataFrame
total_nan_count = city_data_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                                    0
Incident Date                                        0
Incident Time                                        0
Incident Year                                        0
Incident Day of Week                                 0
Report Datetime                                      0
Row ID                                               0
Incident ID                                          0
Incident Number                                      0
CAD Number                                      176929
Report Type Code                                     0
Report Type Description                              0
Incident Code                                        0
Incident Category                                  699
Incident Subcategory                               699
Incident Description                                 0
Resolution                                           0
Intersection                               

In [117]:
# Print column names of city_data_df
print("Column names: ")
print(city_data_df.columns.tolist())

Column names: 
['Incident Datetime', 'Incident Date', 'Incident Time', 'Incident Year', 'Incident Day of Week', 'Report Datetime', 'Row ID', 'Incident ID', 'Incident Number', 'CAD Number', 'Report Type Code', 'Report Type Description', 'Incident Code', 'Incident Category', 'Incident Subcategory', 'Incident Description', 'Resolution', 'Intersection', 'Police District', 'Analysis Neighborhood', 'Supervisor District', 'Latitude', 'Longitude', 'Point', 'Civic Center Harm Reduction Project Boundary', 'HSOC Zones as of 2018-06-05', 'Current Supervisor Districts', 'Current Police Districts']


In [118]:
# Remove rows where 'Police District' is 'Out of SF' in the original city_data_df DataFrame
city_data_df.drop(city_data_df[city_data_df['Police District'] == 'Out of SF'].index, inplace=True)

In [119]:
city_data_df

Incident Datetime Incident Date Incident Time  Incident Year  \
0         10/13/23 23:50      10/13/23   11:50:00 PM           2023   
1         10/13/23 23:50      10/13/23   11:50:00 PM           2023   
2         10/13/23 20:40      10/13/23    8:40:00 PM           2023   
3         10/13/23 20:40      10/13/23    8:40:00 PM           2023   
4         10/13/23 20:37      10/13/23    8:37:00 PM           2023   
...                  ...           ...           ...            ...   
783432       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783433       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783435       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783436       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783437       1/1/18 0:00        1/1/18   12:00:00 AM           2018   

       Incident Day of Week Report Datetime        Row ID  Incident ID  \
0                    Friday  10/13/23 23:50  1.330000e+11      1327919   
1                    Friday  10/13/23 23:50  1.330000e+11      1327919   
2                    Friday  10/13/23 20:49  1.330000e+11      1328218   
3                    Friday  10/13/23 20:49  1.330000e+11      1328218   
4                    Friday  10/13/23 20:49  1.330000e+11      1328221   
...                     ...             ...           ...          ...   
783432               Monday   7/14/20 13:57  9.436317e+10       943631   
783433               Monday  11/20/20 14:21  9.803191e+10       980319   
783435               Monday   5/28/21 12:57  1.030000e+11      1034762   
783436               Monday    5/4/21 18:13  1.030000e+11      1030988   
783437               Monday   8/10/21 13:10  1.060000e+11      1058884   

        Incident Number   CAD Number  ... Police District  \
0             230729452  232853587.0  ...         Bayview   
1             230729452  232853587.0  ...         Bayview   
2             230736946  232863345.0  ...        Southern   
3             230736946  232863345.0  ...        Southern   
4             230736924  232863318.0  ...         Bayview   
...                 ...          ...  ...             ...   
783432        200421296  201961703.0  ...      Tenderloin   
783433        200701703  203251939.0  ...        Northern   
783435        210329436  211481637.0  ...         Taraval   
783436        216055213          NaN  ...        Southern   
783437        210508088  212221621.0  ...      Tenderloin   

             Analysis Neighborhood  Supervisor District   Latitude  \
0            Bayview Hunters Point                 10.0  37.721600   
1            Bayview Hunters Point                 10.0  37.721600   
2                  South of Market                  6.0  37.773769   
3                  South of Market                  6.0  37.773769   
4            Bayview Hunters Point                 10.0  37.729543   
...                            ...                  ...        ...   
783432                  Tenderloin                  6.0  37.778719   
783433                      Marina                  2.0  37.803727   
783435  Oceanview/Merced/Ingleside                 11.0  37.721659   
783436                         NaN                  NaN        NaN   
783437                  Tenderloin                  5.0  37.783933   

         Longitude                                           Point  \
0      -122.390745   POINT (-122.39074534279013 37.72159985216247)   
1      -122.390745   POINT (-122.39074534279013 37.72159985216247)   
2      -122.411610  POINT (-122.41161011784295 37.773769345456756)   
3      -122.411610  POINT (-122.41161011784295 37.773769345456756)   
4      -122.386419   POINT (-122.38641847685011 37.72954271057539)   
...            ...                                             ...   
783432 -122.414741    POINT (-122.4147412230519 37.77871942789032)   
783433 -122.428400   POINT (-122.42840027589088 37.80372712877811)   
783435 -122.467615   POINT (-122.46761532399763 37.7216588070

In [120]:
# Count NaN values in each column city_data_df AFTER REMOVING Out of SF in Police Disctrict column
# city_data_df.drop(city_data_df[city_data_df['Police District'] == 'Out of SF']
nan_count_per_column = city_data_df.isna().sum()

# Count total NaN values in the entire DataFrame
total_nan_count = city_data_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)


NaN count per column:
Incident Datetime                                    0
Incident Date                                        0
Incident Time                                        0
Incident Year                                        0
Incident Day of Week                                 0
Report Datetime                                      0
Row ID                                               0
Incident ID                                          0
Incident Number                                      0
CAD Number                                      162166
Report Type Code                                     0
Report Type Description                              0
Incident Code                                        0
Incident Category                                  680
Incident Subcategory                               680
Incident Description                                 0
Resolution                                           0
Intersection                               

## Computer Aided Dispatch (CAD) column "CAD Number" has 162166 NaN entries
We are going to fill that NaN entries with zero so we can still count Computer Aided Dispatch entries if we want to later.

In [121]:
# Computer Aided Dispatch (CAD) column "CAD Number" with NaN will be filled with "0"
city_data_df['CAD Number'].fillna(0, inplace=True)

## Count NaN values in each column AFTER filling CAD Number NaN entries has been filled with '0'

See how CAD Number column now has zero NaN entries :-)

In [122]:
# Count NaN values in each column AFTER filling CAD Number NaN entries 
# has been filled with '0'
nan_count_per_column = city_data_df.isna().sum()

# Count total NaN values in the entire DataFrame
total_nan_count = city_data_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                                    0
Incident Date                                        0
Incident Time                                        0
Incident Year                                        0
Incident Day of Week                                 0
Report Datetime                                      0
Row ID                                               0
Incident ID                                          0
Incident Number                                      0
CAD Number                                           0
Report Type Code                                     0
Report Type Description                              0
Incident Code                                        0
Incident Category                                  680
Incident Subcategory                               680
Incident Description                                 0
Resolution                                           0
Intersection                               

## We will filter out Incident Category with NaN

This will decrease the dataset by 680 Entries. Caveat: these entries are mostly "complaints" about loud noise and/or incidents that **Do Not** require police presence

In [123]:
# Filter out rows where 'Incident Category' is empty
city_data_df = city_data_df.dropna(subset=['Incident Category'])

In [124]:
# Count NaN values in each column AFTER getting rid of 680 BLANK 'Incident Categories' entries

nan_count_per_column = city_data_df.isna().sum()

# Count total NaN values in the entire DataFrame
total_nan_count = city_data_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                                    0
Incident Date                                        0
Incident Time                                        0
Incident Year                                        0
Incident Day of Week                                 0
Report Datetime                                      0
Row ID                                               0
Incident ID                                          0
Incident Number                                      0
CAD Number                                           0
Report Type Code                                     0
Report Type Description                              0
Incident Code                                        0
Incident Category                                    0
Incident Subcategory                                 0
Incident Description                                 0
Resolution                                           0
Intersection                               

## We are going to fill 'Civic Center Harm Reduction Project Boundary' column entries containing NaN with '0' value. 

This will not affect the over all integrity of the dataset.

In [125]:
city_data_df['Civic Center Harm Reduction Project Boundary'].fillna(0, inplace=True)

/var/folders/br/mc3d1bd90pjblh111j3n3v000000gn/T/ipykernel_7022/999809092.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  city_data_df['Civic Center Harm Reduction Project Boundary'].fillna(0, inplace=True)


In [126]:
# AFTER filling 'Civic Center Harm Reduction Project Boundary' column entries containing NaN with '0' value
nan_count_per_column = city_data_df.isna().sum()

# Count total NaN values in the entire DataFrame
total_nan_count = city_data_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                                    0
Incident Date                                        0
Incident Time                                        0
Incident Year                                        0
Incident Day of Week                                 0
Report Datetime                                      0
Row ID                                               0
Incident ID                                          0
Incident Number                                      0
CAD Number                                           0
Report Type Code                                     0
Report Type Description                              0
Incident Code                                        0
Incident Category                                    0
Incident Subcategory                                 0
Incident Description                                 0
Resolution                                           0
Intersection                               

## We are going to fill 'HSOC Zones as of 2018-06-05' column entries containing NaN with '0' value. 

This will not affect the over all integrity of the dataset.

In [127]:
city_data_df['HSOC Zones as of 2018-06-05'].fillna(0, inplace=True)

/var/folders/br/mc3d1bd90pjblh111j3n3v000000gn/T/ipykernel_7022/3776617433.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  city_data_df['HSOC Zones as of 2018-06-05'].fillna(0, inplace=True)


In [128]:
## AFTER filling 'HSOC Zones as of 2018-06-05' column entries containing NaN with '0' value. 
# Count NaN values in each column
nan_count_per_column = city_data_df.isna().sum()

# Count total NaN values in the entire DataFrame
total_nan_count = city_data_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                                   0
Incident Date                                       0
Incident Time                                       0
Incident Year                                       0
Incident Day of Week                                0
Report Datetime                                     0
Row ID                                              0
Incident ID                                         0
Incident Number                                     0
CAD Number                                          0
Report Type Code                                    0
Report Type Description                             0
Incident Code                                       0
Incident Category                                   0
Incident Subcategory                                0
Incident Description                                0
Resolution                                          0
Intersection                                    26529
Police

## This is the beginning of the REAL clean up 

In [129]:
# DROPPING ['Intersection'] column.
# This will not affect the overall integrity of the dataset
# moving city_data_df dataframe to city_data_wo_intersection_df dataframe

city_data_wo_intersection_df = city_data_df.drop(columns=['Intersection'])

In [130]:
city_data_wo_intersection_df

Incident Datetime Incident Date Incident Time  Incident Year  \
0         10/13/23 23:50      10/13/23   11:50:00 PM           2023   
1         10/13/23 23:50      10/13/23   11:50:00 PM           2023   
2         10/13/23 20:40      10/13/23    8:40:00 PM           2023   
3         10/13/23 20:40      10/13/23    8:40:00 PM           2023   
4         10/13/23 20:37      10/13/23    8:37:00 PM           2023   
...                  ...           ...           ...            ...   
783432       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783433       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783435       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783436       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783437       1/1/18 0:00        1/1/18   12:00:00 AM           2018   

       Incident Day of Week Report Datetime        Row ID  Incident ID  \
0                    Friday  10/13/23 23:50  1.330000e+11      1327919   
1                    Friday  10/13/23 23:50  1.330000e+11      1327919   
2                    Friday  10/13/23 20:49  1.330000e+11      1328218   
3                    Friday  10/13/23 20:49  1.330000e+11      1328218   
4                    Friday  10/13/23 20:49  1.330000e+11      1328221   
...                     ...             ...           ...          ...   
783432               Monday   7/14/20 13:57  9.436317e+10       943631   
783433               Monday  11/20/20 14:21  9.803191e+10       980319   
783435               Monday   5/28/21 12:57  1.030000e+11      1034762   
783436               Monday    5/4/21 18:13  1.030000e+11      1030988   
783437               Monday   8/10/21 13:10  1.060000e+11      1058884   

        Incident Number   CAD Number  ... Police District  \
0             230729452  232853587.0  ...         Bayview   
1             230729452  232853587.0  ...         Bayview   
2             230736946  232863345.0  ...        Southern   
3             230736946  232863345.0  ...        Southern   
4             230736924  232863318.0  ...         Bayview   
...                 ...          ...  ...             ...   
783432        200421296  201961703.0  ...      Tenderloin   
783433        200701703  203251939.0  ...        Northern   
783435        210329436  211481637.0  ...         Taraval   
783436        216055213          0.0  ...        Southern   
783437        210508088  212221621.0  ...      Tenderloin   

             Analysis Neighborhood  Supervisor District   Latitude  \
0            Bayview Hunters Point                 10.0  37.721600   
1            Bayview Hunters Point                 10.0  37.721600   
2                  South of Market                  6.0  37.773769   
3                  South of Market                  6.0  37.773769   
4            Bayview Hunters Point                 10.0  37.729543   
...                            ...                  ...        ...   
783432                  Tenderloin                  6.0  37.778719   
783433                      Marina                  2.0  37.803727   
783435  Oceanview/Merced/Ingleside                 11.0  37.721659   
783436                         NaN                  NaN        NaN   
783437                  Tenderloin                  5.0  37.783933   

         Longitude                                           Point  \
0      -122.390745   POINT (-122.39074534279013 37.72159985216247)   
1      -122.390745   POINT (-122.39074534279013 37.72159985216247)   
2      -122.411610  POINT (-122.41161011784295 37.773769345456756)   
3      -122.411610  POINT (-122.41161011784295 37.773769345456756)   
4      -122.386419   POINT (-122.38641847685011 37.72954271057539)   
...            ...                                             ...   
783432 -122.414741    POINT (-122.4147412230519 37.77871942789032)   
783433 -122.428400   POINT (-122.42840027589088 37.80372712877811)   
783435 -122.467615   POINT (-122.46761532399763 37.7216588070

In [131]:
# Count NaN values in each column city_data_wo_intersection_df
nan_count_per_column = city_data_wo_intersection_df.isna().sum()

# Count total NaN values in the entire DataFrame city_data_wo_intersection_df
total_nan_count = city_data_wo_intersection_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                                   0
Incident Date                                       0
Incident Time                                       0
Incident Year                                       0
Incident Day of Week                                0
Report Datetime                                     0
Row ID                                              0
Incident ID                                         0
Incident Number                                     0
CAD Number                                          0
Report Type Code                                    0
Report Type Description                             0
Incident Code                                       0
Incident Category                                   0
Incident Subcategory                                0
Incident Description                                0
Resolution                                          0
Police District                                     0
Analys

## Save SF_Crime_Data_from_2018_to_Present_wo_Intersection_Oct_18_220pm.csv

In [132]:
city_data_wo_intersection_df.to_csv("SF_Crime_Data_from_2018_to_Present_wo_Intersection_Oct_18_332pm.csv", index=False, encoding="ISO-8859-1")


## Dropping NaN Longtitude and Latitude entries.

This will decrease the original dataset around 5%


In [133]:
city_data_wo_lon_lat_df = city_data_wo_intersection_df.dropna(subset=['Latitude'])

In [134]:
city_data_wo_lon_lat_df.tail()

Incident Datetime Incident Date Incident Time  Incident Year  \
783431       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783432       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783433       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783435       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783437       1/1/18 0:00        1/1/18   12:00:00 AM           2018   

       Incident Day of Week Report Datetime        Row ID  Incident ID  \
783431               Monday   9/29/20 12:29  9.653111e+10       965311   
783432               Monday   7/14/20 13:57  9.436317e+10       943631   
783433               Monday  11/20/20 14:21  9.803191e+10       980319   
783435               Monday   5/28/21 12:57  1.030000e+11      1034762   
783437               Monday   8/10/21 13:10  1.060000e+11      1058884   

        Incident Number   CAD Number  ... Police District  \
783431        200586298  202731493.0  ...        Northern   
783432        200421296  201961703.0  ...      Tenderloin   
783433        200701703  203251939.0  ...        Northern   
783435        210329436  211481637.0  ...         Taraval   
783437        210508088  212221621.0  ...      Tenderloin   

             Analysis Neighborhood  Supervisor District   Latitude  \
783431            Western Addition                  2.0  37.788280   
783432                  Tenderloin                  6.0  37.778719   
783433                      Marina                  2.0  37.803727   
783435  Oceanview/Merced/Ingleside                 11.0  37.721659   
783437                  Tenderloin                  5.0  37.783933   

         Longitude                                          Point  \
783431 -122.423594  POINT (-122.42359407328958 37.78827973195545)   
783432 -122.414741   POINT (-122.4147412230519 37.77871942789032)   
783433 -122.428400  POINT (-122.42840027589088 37.80372712877811)   
783435 -122.467615  POINT (-122.46761532399763 37.72165880704187)   
783437 -122.412595   POINT (-122.41259527758581 37.7839325760642)   

       Civic Center Harm Reduction Project Boundary  \
783431                                          0.0   
783432                                          1.0   
783433                                          0.0   
783435                                          0.0   
783437                                          1.0   

       HSOC Zones as of 2018-06-05 Current Supervisor Districts  \
783431                         0.0                          6.0   
783432                         1.0                         10.0   
783433                         0.0                          6.0   
783435                         0.0                          1.0   
783437                         0.0                         10.0   

        Current Police Districts  
783431                       4.0  
783432                       5.0  
783433                       4.0  
783435                      10.0  
783437                       5.0  

[5 rows x 27 columns]

In [135]:
# Count NaN values in each column city_data_wo_lon_lat_df

nan_count_per_column = city_data_wo_lon_lat_df.isna().sum()

# Count total NaN values in the entire DataFrame city_data_wo_lon_lat_df
total_nan_count = city_data_wo_lon_lat_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                                 0
Incident Date                                     0
Incident Time                                     0
Incident Year                                     0
Incident Day of Week                              0
Report Datetime                                   0
Row ID                                            0
Incident ID                                       0
Incident Number                                   0
CAD Number                                        0
Report Type Code                                  0
Report Type Description                           0
Incident Code                                     0
Incident Category                                 0
Incident Subcategory                              0
Incident Description                              0
Resolution                                        0
Police District                                   0
Analysis Neighborhood                     

In [136]:
# Save dataset to SF_Crime_Data_from_2018_to_Present_wo_Lon_Lat_Oct_18_335pm.csv

city_data_wo_lon_lat_df.to_csv("SF_Crime_Data_from_2018_to_Present_wo_Lon_Lat_Oct_18_335pm.csv", index=False, encoding="ISO-8859-1")

## More clean up. Dropping rows with NaN 'Current Police District'

In [137]:
city_data_wo_cur_pol_disctrict_df = city_data_wo_lon_lat_df.dropna(subset=['Current Police Districts'])

In [138]:
city_data_wo_cur_pol_disctrict_df.tail()

Incident Datetime Incident Date Incident Time  Incident Year  \
783431       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783432       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783433       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783435       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783437       1/1/18 0:00        1/1/18   12:00:00 AM           2018   

       Incident Day of Week Report Datetime        Row ID  Incident ID  \
783431               Monday   9/29/20 12:29  9.653111e+10       965311   
783432               Monday   7/14/20 13:57  9.436317e+10       943631   
783433               Monday  11/20/20 14:21  9.803191e+10       980319   
783435               Monday   5/28/21 12:57  1.030000e+11      1034762   
783437               Monday   8/10/21 13:10  1.060000e+11      1058884   

        Incident Number   CAD Number  ... Police District  \
783431        200586298  202731493.0  ...        Northern   
783432        200421296  201961703.0  ...      Tenderloin   
783433        200701703  203251939.0  ...        Northern   
783435        210329436  211481637.0  ...         Taraval   
783437        210508088  212221621.0  ...      Tenderloin   

             Analysis Neighborhood  Supervisor District   Latitude  \
783431            Western Addition                  2.0  37.788280   
783432                  Tenderloin                  6.0  37.778719   
783433                      Marina                  2.0  37.803727   
783435  Oceanview/Merced/Ingleside                 11.0  37.721659   
783437                  Tenderloin                  5.0  37.783933   

         Longitude                                          Point  \
783431 -122.423594  POINT (-122.42359407328958 37.78827973195545)   
783432 -122.414741   POINT (-122.4147412230519 37.77871942789032)   
783433 -122.428400  POINT (-122.42840027589088 37.80372712877811)   
783435 -122.467615  POINT (-122.46761532399763 37.72165880704187)   
783437 -122.412595   POINT (-122.41259527758581 37.7839325760642)   

       Civic Center Harm Reduction Project Boundary  \
783431                                          0.0   
783432                                          1.0   
783433                                          0.0   
783435                                          0.0   
783437                                          1.0   

       HSOC Zones as of 2018-06-05 Current Supervisor Districts  \
783431                         0.0                          6.0   
783432                         1.0                         10.0   
783433                         0.0                          6.0   
783435                         0.0                          1.0   
783437                         0.0                         10.0   

        Current Police Districts  
783431                       4.0  
783432                       5.0  
783433                       4.0  
783435                      10.0  
783437                       5.0  

[5 rows x 27 columns]

In [139]:
# Count NaN values in each column city_data_wo_cur_pol_disctrict_df
nan_count_per_column = city_data_wo_cur_pol_disctrict_df.isna().sum()

# Count total NaN values in the entire DataFrame city_data_wo_cur_pol_disctrict_df
total_nan_count = city_data_wo_cur_pol_disctrict_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                                 0
Incident Date                                     0
Incident Time                                     0
Incident Year                                     0
Incident Day of Week                              0
Report Datetime                                   0
Row ID                                            0
Incident ID                                       0
Incident Number                                   0
CAD Number                                        0
Report Type Code                                  0
Report Type Description                           0
Incident Code                                     0
Incident Category                                 0
Incident Subcategory                              0
Incident Description                              0
Resolution                                        0
Police District                                   0
Analysis Neighborhood                     

In [140]:
# Save dataset to SF_Crime_Data_from_2018_to_Present_wo_cur_pol_disctrict_df_Oct_18_337pm.csv

city_data_wo_cur_pol_disctrict_df.to_csv("SF_Crime_Data_from_2018_to_Present_wo_cur_pol_disctrict_df_Oct_18_337pm.csv", index=False, encoding="ISO-8859-1")

## Dropping 157 rows with 'Supervisor District' entry of NaN

In [141]:
# Drop 157 rows of NaN 'Supervisor District'

city_data_wo_supervisor_disctrict_df = city_data_wo_cur_pol_disctrict_df.dropna(subset=['Supervisor District'])


In [142]:
city_data_wo_supervisor_disctrict_df.tail()


Incident Datetime Incident Date Incident Time  Incident Year  \
783431       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783432       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783433       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783435       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783437       1/1/18 0:00        1/1/18   12:00:00 AM           2018   

       Incident Day of Week Report Datetime        Row ID  Incident ID  \
783431               Monday   9/29/20 12:29  9.653111e+10       965311   
783432               Monday   7/14/20 13:57  9.436317e+10       943631   
783433               Monday  11/20/20 14:21  9.803191e+10       980319   
783435               Monday   5/28/21 12:57  1.030000e+11      1034762   
783437               Monday   8/10/21 13:10  1.060000e+11      1058884   

        Incident Number   CAD Number  ... Police District  \
783431        200586298  202731493.0  ...        Northern   
783432        200421296  201961703.0  ...      Tenderloin   
783433        200701703  203251939.0  ...        Northern   
783435        210329436  211481637.0  ...         Taraval   
783437        210508088  212221621.0  ...      Tenderloin   

             Analysis Neighborhood  Supervisor District   Latitude  \
783431            Western Addition                  2.0  37.788280   
783432                  Tenderloin                  6.0  37.778719   
783433                      Marina                  2.0  37.803727   
783435  Oceanview/Merced/Ingleside                 11.0  37.721659   
783437                  Tenderloin                  5.0  37.783933   

         Longitude                                          Point  \
783431 -122.423594  POINT (-122.42359407328958 37.78827973195545)   
783432 -122.414741   POINT (-122.4147412230519 37.77871942789032)   
783433 -122.428400  POINT (-122.42840027589088 37.80372712877811)   
783435 -122.467615  POINT (-122.46761532399763 37.72165880704187)   
783437 -122.412595   POINT (-122.41259527758581 37.7839325760642)   

       Civic Center Harm Reduction Project Boundary  \
783431                                          0.0   
783432                                          1.0   
783433                                          0.0   
783435                                          0.0   
783437                                          1.0   

       HSOC Zones as of 2018-06-05 Current Supervisor Districts  \
783431                         0.0                          6.0   
783432                         1.0                         10.0   
783433                         0.0                          6.0   
783435                         0.0                          1.0   
783437                         0.0                         10.0   

        Current Police Districts  
783431                       4.0  
783432                       5.0  
783433                       4.0  
783435                      10.0  
783437                       5.0  

[5 rows x 27 columns]

In [143]:
# Count NaN values in each column city_data_wo_supervisor_disctrict_df
nan_count_per_column = city_data_wo_supervisor_disctrict_df.isna().sum()

# Count total NaN values in the entire DataFrame city_data_wo_cur_pol_disctrict_df
total_nan_count = city_data_wo_supervisor_disctrict_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                               0
Incident Date                                   0
Incident Time                                   0
Incident Year                                   0
Incident Day of Week                            0
Report Datetime                                 0
Row ID                                          0
Incident ID                                     0
Incident Number                                 0
CAD Number                                      0
Report Type Code                                0
Report Type Description                         0
Incident Code                                   0
Incident Category                               0
Incident Subcategory                            0
Incident Description                            0
Resolution                                      0
Police District                                 0
Analysis Neighborhood                           0
Supervisor District         

In [144]:
# This is the final cleaned dataset

city_data_wo_supervisor_disctrict_df.to_csv("SF_Crime_Data_from_2018_to_Presentcity_data_wo_supervisor_disctrict_df_Oct_18_339pm.csv", index=False, encoding="ISO-8859-1")

## Moving city_data_wo_supervisor_disctrict_df to city_data_cleaned_df dataframe

In [145]:
# Create a new DataFrame called 'city_data_cleaned_df' and copy the content of city_data_wo_supervisor_disctrict_df

city_data_cleaned_df = city_data_wo_supervisor_disctrict_df.copy()

In [146]:
city_data_cleaned_df.tail()

Incident Datetime Incident Date Incident Time  Incident Year  \
783431       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783432       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783433       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783435       1/1/18 0:00        1/1/18   12:00:00 AM           2018   
783437       1/1/18 0:00        1/1/18   12:00:00 AM           2018   

       Incident Day of Week Report Datetime        Row ID  Incident ID  \
783431               Monday   9/29/20 12:29  9.653111e+10       965311   
783432               Monday   7/14/20 13:57  9.436317e+10       943631   
783433               Monday  11/20/20 14:21  9.803191e+10       980319   
783435               Monday   5/28/21 12:57  1.030000e+11      1034762   
783437               Monday   8/10/21 13:10  1.060000e+11      1058884   

        Incident Number   CAD Number  ... Police District  \
783431        200586298  202731493.0  ...        Northern   
783432        200421296  201961703.0  ...      Tenderloin   
783433        200701703  203251939.0  ...        Northern   
783435        210329436  211481637.0  ...         Taraval   
783437        210508088  212221621.0  ...      Tenderloin   

             Analysis Neighborhood  Supervisor District   Latitude  \
783431            Western Addition                  2.0  37.788280   
783432                  Tenderloin                  6.0  37.778719   
783433                      Marina                  2.0  37.803727   
783435  Oceanview/Merced/Ingleside                 11.0  37.721659   
783437                  Tenderloin                  5.0  37.783933   

         Longitude                                          Point  \
783431 -122.423594  POINT (-122.42359407328958 37.78827973195545)   
783432 -122.414741   POINT (-122.4147412230519 37.77871942789032)   
783433 -122.428400  POINT (-122.42840027589088 37.80372712877811)   
783435 -122.467615  POINT (-122.46761532399763 37.72165880704187)   
783437 -122.412595   POINT (-122.41259527758581 37.7839325760642)   

       Civic Center Harm Reduction Project Boundary  \
783431                                          0.0   
783432                                          1.0   
783433                                          0.0   
783435                                          0.0   
783437                                          1.0   

       HSOC Zones as of 2018-06-05 Current Supervisor Districts  \
783431                         0.0                          6.0   
783432                         1.0                         10.0   
783433                         0.0                          6.0   
783435                         0.0                          1.0   
783437                         0.0                         10.0   

        Current Police Districts  
783431                       4.0  
783432                       5.0  
783433                       4.0  
783435                      10.0  
783437                       5.0  

[5 rows x 27 columns]

In [147]:
# Count NaN values in each column city_data_cleaned_df
nan_count_per_column = city_data_cleaned_df.isna().sum()

# Count total NaN values in the entire DataFrame city_data_cleaned_df
total_nan_count = city_data_cleaned_df.isna().sum().sum()

print("NaN count per column:")
print(nan_count_per_column)

print("\nTotal NaN count in the DataFrame:", total_nan_count)

NaN count per column:
Incident Datetime                               0
Incident Date                                   0
Incident Time                                   0
Incident Year                                   0
Incident Day of Week                            0
Report Datetime                                 0
Row ID                                          0
Incident ID                                     0
Incident Number                                 0
CAD Number                                      0
Report Type Code                                0
Report Type Description                         0
Incident Code                                   0
Incident Category                               0
Incident Subcategory                            0
Incident Description                            0
Resolution                                      0
Police District                                 0
Analysis Neighborhood                           0
Supervisor District         

## Save cleaned dataset to SF_Crime_Data_from_2018_to_Present_city_data_cleaned_df_Oct_18_345pm.csv file

In [148]:
# Save SF_Crime_Data_from_2018_to_Present_city_data_cleaned_df_Oct_18_345pm.csv

city_data_cleaned_df.to_csv("SF_Crime_Data_from_2018_to_Present_city_data_cleaned_df_Oct_18_345pm.csv", index=False, encoding="ISO-8859-1")

In [149]:
# Save SF_Crime_Data_from_2018_to_Present_CLEANED.csv

city_data_cleaned_df.to_csv("SF_Crime_Data_from_2018_to_Present_CLEANED.csv", index=False, encoding="ISO-8859-1")

## The DropBox location file of "SF_Crime_Data_from_2018_to_Present_CLEANED.csv" 

https://www.dropbox.com/scl/fi/f28k1ztt43nwuc4elzf44/SF_Crime_Data_from_2018_to_Present_CLEANED.csv?rlkey=bya3p7s6idl3hmwx48c13stux&dl=0